python/ray/rllib/ppo/ppo_evaluator.py

```diff
@@ -132,45 +84,64 @@ def build_loss(obs, vtargets, advs, acts, plog, pvf_preds):
         self.sampler = SyncSampler(
             self.env, self.common_policy, self.obs_filter,
             self.config["horizon"], self.config["horizon"])
-        self.sess.run(tf.global_variables_initializer())
-
-    def load_data(self, trajectories, full_trace):
-        use_gae = self.config["use_gae"]
-        dummy = np.zeros_like(trajectories["advantages"])
-        return self.par_opt.load_data(
-            self.sess,
-            [trajectories["obs"],
-             trajectories["value_targets"] if use_gae else dummy,
-             trajectories["advantages"],
-             trajectories["actions"],
-             trajectories["logprobs"],
-             trajectories["vf_preds"] if use_gae else dummy],
-            full_trace=full_trace)
-
-    def run_sgd_minibatch(
-            self, batch_index, kl_coeff, full_trace, file_writer):
-        return self.par_opt.optimize(
-            self.sess,
-            batch_index,
-            extra_ops=[
-                self.mean_loss, self.mean_policy_loss, self.mean_vf_loss,
-                self.mean_kl, self.mean_entropy],
-            extra_feed_dict={self.kl_coeff: kl_coeff},
-            file_writer=file_writer if full_trace else None)
-
-    def compute_gradients(self, samples):
-        raise NotImplementedError
-
-    def apply_gradients(self, grads):
-        raise NotImplementedError
+
+    def tf_loss_inputs(self):
+        return self.inputs
+
+    def build_tf_loss(self, input_placeholders):
+        obs, vtargets, advs, acts, plog, pvf_preds = input_placeholders
+        return ProximalPolicyGraph(
+            self.env.observation_space, self.env.action_space,
+            obs, vtargets, advs, acts, plog, pvf_preds, self.logit_dim,
+            self.kl_coeff, self.distribution_class, self.config,
+            self.sess, self.registry)
+
+    def init_extra_ops(self, device_losses):
```

Comment:
This was odd because of the chain of dependencies:

1. Local evaluator creates a model
2. MultiGPU creates Variable replicas (which are just refs to the local model)
3. These ops are created, which use nodes from (2)